app/models/activities/steps_management.rb

```diff
@@ -17,15 +18,16 @@ def steps_for(assets)
   end
 
   def step_types_for(assets, required_assets = nil)
-    stypes = step_types.not_for_reasoning.includes(:condition_groups => :conditions).select do |step_type|
+    # @todo This could be moved to the relationship
+    compatible_step_types = step_types.not_for_reasoning.includes(condition_groups: :conditions).distinct.select do |step_type|
       step_type.compatible_with?(assets, required_assets)
     end.uniq
-    stype = stypes.detect { |stype| steps.in_progress.for_step_type(stype).count > 0 }
-    stype.nil? ? stypes : [stype]
+    in_progress_step_type_ids = steps.in_progress.group(:step_type_id).pluck(:step_type_id)
+    in_progress_step_type = compatible_step_types.detect { |stype| in_progress_step_type_ids.include?(stype.id) }
+    in_progress_step_type.nil? ? compatible_step_types : [in_progress_step_type]
```

Comment:
Here was mostly a case of:
1) Trying to improve local variable names
2) Avoid mutating local variables
3) Reduce the number of hits to the database by pulling in in-ptogress steps outside the loop.